### PR TITLE
MAINT: Improve speed of ufunc kwargs parsing

### PR DESCRIPTION
**Pull request by mhvk**
This is an alternative to #11333, which doesn't use the enum and includes a drastic rewrite in terms of  conversion functions instead of the enormous switch/case loop (except for "out"). The reductions in overhead are the same as with #11333 to within a few ns (see below).

@eric-wieser - what do you think? My own sense is that this will be much more easily used also for `reduce`, etc., as it follows the python parsing ideas more closely.

EDIT: added reductions in overhead from #11333, since I closed that one.
```
import numpy as np
a = np.array(1)
b = np.empty_like(a)
%timeit np.positive(a)                                 # 352->348 ns
%timeit np.positive(a, subok=True)                     # 606->501 ns
%timeit np.positive(a, where=True)                     # 586->503 ns
%timeit np.positive(a, where=True, subok=True)         # 695->531 ns
%timeit np.positive(a, b)                              # 354->352 ns
%timeit np.positive(a, out=b)                          # 557->480 ns
%timeit np.positive(a, out=b, subok=True)              # 668->506 ns
%timeit np.positive(a, out=b, where=True, subok=True)  # 752->536 ns
```

### CODE REVIEW

File: numpy/core/src/umath/ufunc_object.c
Comment (eric-wieser) [inline numpy/core/src/umath/ufunc_object.c]: This block isn't threadsafe

File: numpy/core/src/umath/ufunc_object.c
Comment (eric-wieser) [inline numpy/core/src/umath/ufunc_object.c]: Worth adding that this is because the strings are not interned yet during static initialization.

You could store `&npy_um_str_out` etc into the array instead, which would work fine during static initialization

File: numpy/core/src/umath/ufunc_object.c
Comment (eric-wieser) [inline numpy/core/src/umath/ufunc_object.c]: I often like to do my function typedefs as
```
typedef int converter(PyObject *, void *)
```

And then use `converter *convert` below, which reads as "a pointer to a converter function"

**Comment by eric-wieser** [comment]
I wonder if you should just make a `va_list` function, which you could call as

```
parse_kwargs_fast(kwargs,
    interned_name, parser, target_var,
    interned_name, parser, target_var,
    ...
    NULL)
```

I know very little about `va_list`, but my understanding is that it's really just a bunch of helpers to walk through the stack

File: numpy/core/src/umath/ufunc_object.c
Comment (mhvk) [reply numpy/core/src/umath/ufunc_object.c]: I had mine from the CPython implementation, but that definitely looks nicer, so I changed it.

File: numpy/core/src/umath/ufunc_object.c
Comment (mhvk) [reply numpy/core/src/umath/ufunc_object.c]: OK, better indeed.

File: numpy/core/src/umath/ufunc_object.c
Comment (mhvk) [reply numpy/core/src/umath/ufunc_object.c]: Oops, I had not thought of that... I actually had a parallel array before, and at the time it seemed less nice. But reverting, I actually think it is much nicer, so I adopted it.

**Comment by mhvk** [comment]
@eric-wieser - I followed your suggestions and think it looks quite a bit better still. Do you agree that this is a better solution than #11333? If so, I'll close that one.

On the `va_list`, I did think of making effectively a sped-up version of `PyArg_ParseArgsAndKeywords`, but that may be better done in a separate PR. Also, I think for this routine, we may want to do `out` before it even gets called (since it is needed even if the ufunc is overridden).

**Comment by mattip** [comment]
I see #113333 is still open, but this is the desired alternative? Can we merge this and close the other?

**Comment by mhvk** [comment]
I definitely think this is the better solution -- just wasn't sure whether there were further comments.

File: numpy/core/src/umath/ufunc_object.c
Comment (mhvk) [inline numpy/core/src/umath/ufunc_object.c]: This could almost become a fast keyword parsing function that we could use also for the reductions. But as is, this part of code is not able to `DECREF` items on failure, so I'd rather leave that for another PR (it needs changing the converters such that if passed `NULL`, they decref if needed).

File: numpy/core/src/umath/ufunc_object.c
Comment (mhvk) [inline numpy/core/src/umath/ufunc_object.c]: I'm still not completely sold on this new struct - its main advantage of keeping the names and converters together perhaps doesn't outweigh adding it -- but for use in just one function it is probably fine (and another reason for not yet aiming to make a general keyword parsing function).

**Comment by mhvk** [comment]
Rebased, and refactored a little more. There are some obvious further improvements possible, but I'd prefer to leave those for a future PR (see in-line comment).

**Comment by mhvk** [comment]
@eric-wieser - sorry to bug for two PRs, but this one would also be nice to get in...

File: numpy/core/src/umath/ufunc_object.c
Comment (eric-wieser) [inline numpy/core/src/umath/ufunc_object.c]: This won#y be true once umath and multiarray are merged, right?

File: numpy/core/src/umath/ufunc_object.c
Comment (eric-wieser) [inline numpy/core/src/umath/ufunc_object.c]: Having this order duplicated between here and above seems very fragile. I'll see if I can think of another way

File: numpy/core/src/umath/ufunc_object.c
Comment (eric-wieser) [inline numpy/core/src/umath/ufunc_object.c]: FYI: No need to dereference the function pointer here - `*fp()` and `fp()` [are synonymous in C](https://stackoverflow.com/q/7518815/102441). I'm fine with this line as is though

File: numpy/core/src/umath/ufunc_object.c
Comment (eric-wieser) [inline numpy/core/src/umath/ufunc_object.c]: Is special casing `NULL` here really better than just having one more function pointer?

File: numpy/core/src/umath/ufunc_object.c
Comment (eric-wieser) [inline numpy/core/src/umath/ufunc_object.c]: Wait, CPython3x format strings have a special syntax for unicode? I think there's a lot of places that cleanup could go

File: numpy/core/src/umath/ufunc_object.c
Comment (eric-wieser) [inline numpy/core/src/umath/ufunc_object.c]: Can you extract this line to line 787 as a helper function, `ufunc_kwarg *find_ufunc_kwd(PyObject *key)`, which returns null if nothing is found?

That would make it easier to swap it for a pointer-ordered binary search in future, if necessary

File: numpy/core/src/umath/ufunc_object.h
Comment (eric-wieser) [inline numpy/core/src/umath/ufunc_object.h]: Can you add `from umathmodule.c` to this comment?

File: numpy/core/src/umath/ufunc_object.c
Comment (mhvk) [reply numpy/core/src/umath/ufunc_object.c]: Done.

**Comment by mhvk** [comment]
@eric-wieser - you're requests made me realize that, really, things would be much nicer if I just bit the bullet and wrote a parser...  So, that is now done, using a `va_list`. It is somewhat similar the `PyArg_ParseArgsAndKeywords`, though with a list of names in objects instead of in char (and pointers to those instead of the keywords themselves).

Maybe good to first have a quick look; some comments in-line.

File: numpy/core/src/umath/ufunc_object.c
Comment (mhvk) [inline numpy/core/src/umath/ufunc_object.c]: I wonder if I should instead do
```
PyObject *kwnames[13] = {NULL};
```
and then explicitly fill on first call, to get rid of the extra indirections every else.

File: numpy/core/src/umath/ufunc_object.c
Comment (mhvk) [inline numpy/core/src/umath/ufunc_object.c]: I'm not 100% sure this way of repeatedly going through the `va_list` is OK generically, but I guess the tests will tell.

File: numpy/core/src/umath/ufunc_object.c
Comment (mhvk) [inline numpy/core/src/umath/ufunc_object.c]: Obviously, some repetition here with the above. Not quite sure how to solve it.

**Comment by mattip** [comment]
Maybe I missed it, but did you try replacing this with `PyArg_ParseArgsAndKeywords` ? It seems you are moving in the direction of duplicating much of that code

**Comment by mhvk** [comment]
@mattip - the problem is that `PyArg_ParseArgsAndKeywords` is very slow, in part because it turns every character keyword into an object, in part (IIRC) because it searches for keywords in the dict instead of scanning over the dict. But, yes, effectively we're emulating CPython code (the parts that deal with function calls); sadly, that is not exposed.

**Comment by eric-wieser** [comment]
@mhvk: Do we have a benchmark that proves that?

**Comment by eric-wieser** [comment]
Can you add a benchmark that compares:
```
np.add(a, b, out=c)
np.add(a, b, c)
np.add(a, b, out=(c,))
```
for some small arrays? Preferably in some way that `asv` will show them all on the same graph

**Comment by mhvk** [comment]
I pushed a version that has the benchmarks as the first commit. I tried installing `asv` and it did run with the simple `./runtest.py --bench` command, but nothing else in `bench/README.rst` just works (no comparisons, no publishing). I don't really have time to debug this, though, and would prefer to focus on whether the PR itself makes sense (I feel it is an improvement over the previous parsing even ignoring the speed gain!).

File: benchmarks/benchmarks/bench_ufunc.py
Comment (eric-wieser) [inline benchmarks/benchmarks/bench_ufunc.py]: I'd like these in a separate PR, simply because the hosted asv runs nightly, so would not produce a before / after if they get merged together.

File: numpy/core/src/umath/ufunc_object.c
Comment (eric-wieser) [inline numpy/core/src/umath/ufunc_object.c]: You'll get a system error here if `convert` tries to do any python stuff since there's an exception in flight. Possibly worth `Fetch`ing the exception, and checking the return value

File: numpy/core/src/umath/ufunc_object.c
Comment (eric-wieser) [inline numpy/core/src/umath/ufunc_object.c]: Can you move the definition of `kwnames` to here?, so that the two things that have to be in the same order are at least adjacent?

File: numpy/core/src/umath/ufunc_object.c
Comment (eric-wieser) [reply numpy/core/src/umath/ufunc_object.c]: Only if profiling shows it's worth it. I'd leave that to a future PR.

File: numpy/core/src/umath/ufunc_object.c
Comment (eric-wieser) [inline numpy/core/src/umath/ufunc_object.c]: Is this really worth it vs just passing these functions directly into the `parse_ufunc_keywords` call below?

File: numpy/core/src/umath/ufunc_object.c
Comment (eric-wieser) [inline numpy/core/src/umath/ufunc_object.c]: Is this necessary now, or does `parse_ufunc_keywords` handle that?

File: numpy/core/src/umath/ufunc_object.c
Comment (eric-wieser) [reply numpy/core/src/umath/ufunc_object.c]: Looks like [you're fine](http://www.cplusplus.com/reference/cstdarg/va_start/):

> `void va_start (va_list ap, paramN);`
> If ap has already been passed as first argument to a previous call to va_start or va_copy, **it shall be passed to va_end before calling this function.**

File: numpy/core/src/umath/ufunc_object.c
Comment (eric-wieser) [reply numpy/core/src/umath/ufunc_object.c]: Could encapsulate the loop::
```
for (i = 0; i <= index; i++) {
    convert = va_arg(va, converter *);
    output = va_arg(va, void *);
}
```
But arguably not worth it.

File: numpy/core/src/umath/ufunc_object.c
Comment (eric-wieser) [inline numpy/core/src/umath/ufunc_object.c]: This condition is never hit.

File: numpy/core/src/umath/ufunc_object.c
Comment (eric-wieser) [inline numpy/core/src/umath/ufunc_object.c]: Not sure this is safe if `obj == NULL`

File: benchmarks/benchmarks/bench_ufunc.py
Comment (mhvk) [reply benchmarks/benchmarks/bench_ufunc.py]: Good point, see #11453

File: numpy/core/src/umath/ufunc_object.c
Comment (mhvk) [reply numpy/core/src/umath/ufunc_object.c]: Duh, yes, of course that should work! Indeed, much clearer, especially with the keywords there too.

File: numpy/core/src/umath/ufunc_object.c
Comment (mhvk) [reply numpy/core/src/umath/ufunc_object.c]: Yes, makes much more sense.

File: numpy/core/src/umath/ufunc_object.c
Comment (mhvk) [reply numpy/core/src/umath/ufunc_object.c]: Yes, still necessary. The parser only touches parts that are actually in `kwds`

File: numpy/core/src/umath/ufunc_object.c
Comment (mhvk) [reply numpy/core/src/umath/ufunc_object.c]: It is if the parser is calling the function again in the failure path to undo an earlier initialization, i.e., in this case when, say, the first keyword was `where` and the second keyword something that is not allowed.

This behaviour is the same as for `PyArg_ParseArgsAndKeywords`

File: numpy/core/src/umath/ufunc_object.c
Comment (mhvk) [reply numpy/core/src/umath/ufunc_object.c]: I thought surely it was but looking at the [actual definition](https://github.com/python/cpython/blob/6f0eb93183519024cb360162bdd81b9faec97ba6/Include/boolobject.h#L12), it clearly is not.

File: numpy/core/src/umath/ufunc_object.c
Comment (mhvk) [reply numpy/core/src/umath/ufunc_object.c]: Indeed, I rewrote as
```
static *kwnames[13] = {NULL}
if (kwnames[0] == NULL) {
    kwnames[0] = npy_um_...
...
}
```
so that I don't have the triple `***kwnames` any more. Should be just that little bit faster.

File: numpy/core/src/umath/ufunc_object.c
Comment (mhvk) [reply numpy/core/src/umath/ufunc_object.c]: p.s. Added a note about the converters to make this clear.

Comment (mhvk) [reply numpy/core/src/umath/ufunc_object.c]: But looking more closely at the array ones, I'm not so sure they are actually written to follow the logic... And indeed, reading the documentation [1] more carefully, a function has to return `Py_CLEANUP_SUPPORTED` and this only works in python 3. I'll rewrite accordingly.

[1] https://docs.python.org/3/c-api/arg.html

**Comment by mhvk** [comment]
@eric-wieser - thanks for the (as usual) very good comments. Now having looked more at the normal parsers, it is clear that the cleanup is better not done inside it; being a bit more judicious with borrowed and new references allowed the code to become substantially simpler. I'm now sufficiently convinced this is the right path that I squashed all commits. In a follow-up commit, I'll use this parser for the reduction methods (which may need a bit of tinkering with the keyword-not-found handler, but should otherwise be very straightforward).

File: numpy/core/src/umath/ufunc_object.c
Comment (eric-wieser) [reply numpy/core/src/umath/ufunc_object.c]: What I meant here is that you had your ifs the wrong way around.

```C
if (obj != PyTrue) {
}
else if (obj == NULL) {
    // but this can never happen, because PyTrue != NULL
}
```

**Review by eric-wieser — APPROVED** [review]
Looks pretty good now - thanks for showing me a few nifty py3 c api features, even if we can't use them yet...

**Comment by eric-wieser** [comment]
Let's not merge this until the benchmark suite picks up your new benchmarks

File: numpy/core/src/umath/ufunc_object.c
Comment (mhvk) [reply numpy/core/src/umath/ufunc_object.c]: Ah, yes, that was silly!

**Comment by eric-wieser** [comment]
Alright, let's see how the benchmarks improve

**Comment by mhvk** [comment]
Great!

**Comment by eric-wieser** [comment]
Doesn't seem to have made any real difference

image

**Pull request merged**
